app/controllers/api/v1/digital_objects_controller.rb

```diff
@@ -22,12 +23,16 @@ def show
       # POST /digital_objects
       # POST /digital_objects.json
       def create
-        digital_object_data = JSON.parse(create_or_update_params['digital_object_data_json'])
-        @digital_object = DigitalObject::Base.class_for_type(digital_object_data['digital_object_type']).new
-        @digital_object.set_digital_object_data(digital_object_data)
-
-        if @digital_object.save
-          render :show, status: :created, location: @digital_object
+        digital_object_data = create_or_update_params
+        digital_object_record = DigitalObjectRecord.create
+        @digital_object = DigitalObject::Base.from_serialized_form(digital_object_record, digital_object_data)
+        @digital_object.projects.each do |project|
+          authorize! :create_objects, project
+        end
+        @digital_object.state ||= Hyacinth::DigitalObject::State::ACTIVE
+        # digital_object_record.save first because it will assign necessary digital_object_record attributes
+        if @digital_object.save && (digital_object_record.persisted? || digital_object_record.save)
+          show
         else
           render json: @digital_object.errors, status: :unprocessable_entity
```

Comment:
I've been wrapping these in an `errors` function that's defined in `ApplicationApiController`. Though the digital object might have us re-think how we display the errors in the json. We can save this for later.